src/ethereum/write_oz/openzeppelin.rs

```diff
@@ -0,0 +1,325 @@
+            error!(?status, ?e, "error occurred, unknown response format");
+            Error::UnknownResponseFormat
+        })?;
+
+        Ok(item)
+    }
+
+    async fn list_transactions(&self) -> Result<Vec<SubmittedTransaction>, Error> {
+        let client = get_client(&self.api_key, &self.api_secret)
+            .await
+            .map_err(|_| Error::Authentication)?;
+
+        let res = client
+            .get(RELAY_TXS_URL)
+            .send()
+            .await
+            .map_err(|_| Error::RequestFailed)?;
+
+        let status = res.status();
+        let items = res.json::<Vec<SubmittedTransaction>>().await.map_err(|e| {
+            error!(?status, ?e, "error occurred, unknown response format");
+            Error::UnknownResponseFormat
+        })?;
+
+        Ok(items)
+    }
+
+    async fn mine_transaction_id(&self, id: &str) -> Result<SubmittedTransaction, TxError> {
+        loop {
+            let transaction = self.query(id).await.map_err(|error| {
+                error!(?error, "Failed to get transaction status");
+                TxError::Send(Box::new(error))
+            })?;
+            let status = transaction
+                .status
+                .as_ref()
+                .ok_or_else(|| TxError::Dropped(TxHash::default()))?;
+
+            if status == "mined" || status == "confirmed" || status == "failed" {
+                return Ok(transaction);
+            }
+
+            info!("waiting 5 s to mine");
+            tokio::time::sleep(Duration::from_secs(5)).await;
+        }
+    }
+
+    async fn send_oz_transaction<T: Into<TypedTransaction> + Send + Sync>(
+        &self,
+        tx: T,
+    ) -> Result<String, Error> {
+        let client = get_client(&self.api_key, &self.api_secret)
+            .await
+            .map_err(|_| Error::Authentication)?;
+
+        let tx: TypedTransaction = tx.into();
+        let api_tx = Transaction {
+            to:          tx.to(),
+            value:       tx.value(),
+            gas_limit:   tx.gas(),
+            data:        tx.data(),
+            valid_until: Some(chrono::Utc::now() + self.transaction_validity),
+        };
+
+        let res = client
+            .post(RELAY_TXS_URL)
+            .body(json!(api_tx).to_string())
+            .send()
+            .await
+            .map_err(|_| Error::RequestFailed)?;
+
+        if res.status() == StatusCode::OK {
+            let obj = res
+                .json::<Value>()
+                .await
+                .map_err(|_| Error::UnknownResponseFormat)?;
+            let id = obj
+                .get("transactionId")
+                .ok_or(Error::MissingTransactionId)?
+                .as_str()
+                .unwrap();
+            Ok(id.to_string())
+        } else {
+            let text = res.text().await;
+            info!(?text, "response error");
+
+            Err(Error::UnknownResponseFormat)
+        }
+    }
+
+    pub async fn send_transaction(
```

Comment:
Do I understand correctly this is very-idempotent? I.e. that I can blindly call this 7 times and be quite certain it will only submit once? This is awesome if true.

Reply:
Sort of. It's... idempotent as long as you set `is_retry = true` and your transaction is in the `list_transactions` history.

According to OZ support, `list_transactions` will return up to 1MB of data (so older transactions get evicted). Also, it might take some time for your transaction to show up on the list, but the support really didn't want to commit to any number:

> To be safe I'll say seconds because there is some uncertainty as to how quickly different elements of our infrastructure will trigger but could be as short as milliseconds.

I should document all of that.

Reply:
Awesome, sounds good

Reply:
Renamed `is_retry` to `only_once`, limited the number of results (our process is very sequential anyway), and added some docs.